### ISSUE TITLE
More dimensions

### ISSUE THREAD

**Issue by jagot** [JuliaApproximation/ContinuumArrays.jl#15]
This is a follow-up to #14, but more speculatory.

At the moment, I'm working with atoms with spherical symmetry, which allows me to reduce my differential equations to one-dimensional radial problems, with the angular coordinates treated exactly using angular momentum algebra. In the future, I will probably want/need to go to molecules which possess less/other symmetries. There are many different ways of treating molecules, but the dominating by far is linear combinations of (multi-centred) [3d Gaussians](https://en.wikipedia.org/wiki/Gaussian_orbital) (actually polynomials multiplied by Gaussians). It would be interesting to see if we could fit these type of basis functions within the ContinuumArrays framework. If we get this to work, a Hartree–Fock code could conceivably be used for both atoms and molecules without modifications (some tensor magic would need to happen).

I am by no means well-versed on the topic, and would need to read up quite a bit (although I have a lot of good references). @jarvist could possibly be interested in this as well.

**Comment by dlfivefifty** [comment]
Yes this needs to be fixed in QuasiArrays.jl. Essentially everywhere that specialises on `::Real` is a placeholder for a quasi-array to support general axes types (that is, we want to dispatch on `eltype.(axes(A))`. Some of these restrictions may be dropped completely, however, we need to be careful with `getindex` which does different things for vector arguments.

**Comment by jagot** [comment]
If I understand it correctly, the indexing you're referring to is to evaluate the approximated function at various coordinates (which is certainly necessary). However, I have understood the philosophy of basis sets such as Gaussians to mainly be the transformation of a continuous problem to a matrix problem (with emphasis on this); sort of the opposite what I feel we're trying to accomplish with ContinuumArrays.jl.

I will try to clarify: with molecules, you have no hope to cover all of `3N`-dimensional space (`N` being the amount of electrons) with the same level of accuracy everywhere. Instead, you place a bunch of Gaussians on top of every atom in your molecule and expand your functions using these. There are all kinds of tricks to optimize (reduce) the basis sets for given molecular structures.

Quantum chemists call this "basis set calculations", and you have to derive your equations in terms of these, rather than the actual functions you're approximating, since they do not obey simple things like completeness relations (`sum |B_i><B_i| != 1`) even when you formally go to infinite amount of basis functions, as do e.g. finite-differences. This is why I'm wondering if we could shoe-horn this into QuasiArrays.jl/ContinuumArrays.jl.

I think my crude explanation is neither fully correct, nor very elucidating, apologies for this.

**Comment by dlfivefifty** [comment]
Is this basically radial basis functions (RBFs)? I think that fit's in naturally.

**Comment by jagot** [comment]
Yes it is.

**Comment by dlfivefifty** [comment]
In ApproxFun.jl I handled this by interlacing the coefficients by blocks: interlacing rows/columns of banded matrices results in a banded matrix, as does banded-block-banded matrices. I started playing with adding this to LazyArrays.jl as a lazy `Interlace` array: 

https://github.com/JuliaArrays/LazyArrays.jl/pull/41/

But in the context of quasi-arrays, some more thought is needed for both concatenation and interlacing. For concatenation, there are multiple options for the `axes(A)` where `A = Vcat(a,b)`: 

1. `axes(A,1) == OneTo(sum(length.(A.args)))`. This is the standard array choice but obviously doesn't work for infinite axes.
2. `axes(A,1) == Inclusion(union(axes(a,1).domain, axes(b,1).domain))` (assuming `a` and `b` have `Inclusion` axes). This is equivalent to `PiecewiseSpace` in ApproxFun. It's not clear what to do where the axes intersect. 
3. `axes(A,1) == axes(a,1) ⊗ OneTo(2)`, (assuming `axes(a,1) === axes(b,2)`). This is equivalent to `ArraySpace` in ApproxFun. One would access entries as `A[SVector(0.1,1)]`

**Comment by dlfivefifty** [comment]
PS For completeness, the analogue of `SumSpace` is `Hcat(a, b)`.

**Comment by jagot** [comment]
I see. It seems 1) is basically what I did above, 2) will not work in this case (I think?), and 3) is the most general one, and which is where I want to go.

An operator `O` acting on the expansion coefficients of such a system will have `axes(O) == (axes(a,1) ⊗ OneTo(2),axes(a,1) ⊗ OneTo(2))`, I guess? I.e. it is a rank-4 tensor.

**Comment by jagot** [comment]
I was reading a bit on how to implement the full time-dependent Schrödinger equation for helium, which has two electrons, corresponding to 6-dimensional space. This quickly becomes intractable with growing sizes of the axes, but the nature of the problem is that the electrons are usually never "far away" from the origin simultaneously. It would be cool if we could somehow fit in "Hyberpolic Reduced Tensor-Product Bases", as described by e.g. [Lubich (2008)](https://www.researchgate.net/publication/237298462_From_Quantum_to_Classical_Molecular_Dynamics_Reduced_Models_and_Numerical_Analysis):

<img width="1281" alt="image" src="https://user-images.githubusercontent.com/117947/69451836-9d948800-0d60-11ea-8504-4d8851be1ca3.png">

**Comment by dlfivefifty** [comment]
This sounds like it’s about tensor Fourier. It wouldn’t be relevant for any basis built from spherical harmonics.

In terms of how it fits in, it’s all about different ways of turning a matrix into a vector. In ApproxFun this was achieved by different ` tensorizer ` types, but it was a bit of a hack. We’ll have to address this when support for multiple dimensions is added, hopefully it will be as easy as adding a special tensirizer type

**Comment by jagot** [comment]
> This sounds like it’s about tensor Fourier. It wouldn’t be relevant for any basis built from spherical harmonics.

This is for the radial coordinates of the two electrons, i.e. `r_1` and `r_2`, where the full tensor product basis becomes too large. The angular coordinates, `θ_1`, `θ_2`, and `φ_12` are treated using bipolar spherical harmonics, they present no problem.
